File: Engine.Core/Events/Event.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace Engine.Core.Events
{
    public class Event
    {
        private delegate void Method();

        private Method collection;

        public void AddListener(Action method)
        {
            Method m = new Method(method);
            collection += m;
        }
        public void RemoveListener(Action method)
        {
            Method m = new Method(method);
            collection -= m;
        }
        public void RemoveAllListeners()
        {
            Delegate.RemoveAll(collection, collection);
        }

        public void Invoke()
        {
            collection?.Invoke();
        }

        public static Event operator +(Event e, Action method)
        {
            e.AddListener(method);
            return e;
        }
        public static Event operator -(Event e, Action method)
        {
            e.RemoveListener(method);
            return e;
        }
    }
}

File: Engine/src/Core/Events/Event.cpp
#include "Event.h"

namespace Engine::Core::Events
{
    void Event::AddListener(const std::function<void()> &method)
    {
        collection.push_back(method);
    }

    void Event::RemoveListener(const std::function<void()> &method)
    {
        collection.erase(std::remove_if(collection.begin(), collection.end(),
                                        [&](const std::function<void()> &storedMethod)
                                        {
                                            return storedMethod.target<void()>() == method.target<void()>();
                                        }),
                         collection.end());
    }

    void Event::RemoveAllListeners()
    {
        collection.clear();
    }

    void Event::Invoke()
    {
        for (const auto &method : collection)
        {
            method();
        }
    }

    Event &Event::operator+=(const std::function<void()> &method)
    {
        AddListener(method);
        return *this;
    }

    Event &Event::operator-=(const std::function<void()> &method)
    {
        RemoveListener(method);
        return *this;
    }
}